Update networkx documentation (#622)
memgraph/docs: docs/how-to-guides/networkx.md

---

NetworkX is a Python package for the creation, manipulation, and study of the structure,
dynamics, and functions of complex networks. Memgraph has [**`nxalg`**](/docs/mage/query-modules/python/nxalg) query module, which is a wrapper around NetworkX graph algorithms. It also provides **[Graph Analyzer](/mage/query-modules/python/graph-analyzer)** query module, which utilizes the NetworkX library. Besides that, you can create a custom query module which uses the NetworkX library. Through this how-to guide, you are going to learn:

- [**How to run NetworkX algorithms**](#how-to-run-networkx-algorithms)
- [**How to implement custom NetworkX module**](#how-to-implement-custom-networkx-module)

> ## How to run NetworkX algorithms

Reviewer: ```suggestion
## How to run NetworkX algorithms in Memgraph Lab
```

> NetworkX algorithms in Memgraph are query modules like any other [MAGE](/docs/mage) query module. Head over to the guide on [how to call MAGE procedures](/docs/mage/usage/calling-procedures) to learn how to call all Memgraph procedures, including those that utilize the NetworkX library.

Reviewer: ```suggestion
NetworkX algorithms are integrated into Memgraph as query modules inside Memgraph’s open-source graph extension library [MAGE](/docs/mage). Head over to the guide on [how to call MAGE procedures](/docs/mage/usage/calling-procedures) to learn how to call all Memgraph procedures, including those that utilize the NetworkX library.
```

> In this how-to guide, we will show one simple example of how to call a NetworkX procedure. The prerequisite of following the example is to have a running Memgraph instance. To run Memgraph with the visual user interface, use the Memgraph Platform Docker image where Memgraph Lab is included. Run the following command:

Reviewer: ```suggestion
This how-to guide will show one simple example of calling a NetworkX procedure. 
To be able to try out this example on your own, you need a running Memgraph instance. Run Memgraph using the Memgraph Platform Docker image, which includes both the MAGE library and Memgraph Lab, a visual interface. 

To run the image, open a command-line interpreter and run the following Docker command:
```

```
docker run -it -p 7687:7687 -p 7444:7444 -p 3000:3000 memgraph/memgraph-platform:latest
```

> Check out the [installation guide](/docs/memgraph/installation) for other installation options.

Reviewer: ```suggestion
For other installation options, check out the [installation guide](/docs/memgraph/installation).
```

---

NetworkX is a Python package for the creation, manipulation, and study of the structure,

> dynamics, and functions of complex networks. Memgraph has [**`nxalg`**](/docs/mage/query-modules/python/nxalg) query module, which is a wrapper around NetworkX graph algorithms. It also provides **[Graph Analyzer](/mage/query-modules/python/graph-analyzer)** query module, which utilizes the NetworkX library. Besides that, you can create a custom query module which uses the NetworkX library. Through this how-to guide, you are going to learn:

Reviewer: ```suggestion
dynamics, and functions of complex networks. Memgraph has [**`nxalg`**](/docs/mage/query-modules/python/nxalg) query module, which is a wrapper around NetworkX graph algorithms. It also provides **[Graph Analyzer](/mage/query-modules/python/graph-analyzer)** query module, which utilizes the NetworkX library. Besides that, you can create a custom query module that uses the NetworkX library. Through this how-to guide, you will learn:
```

- [**How to run NetworkX algorithms in Memgraph Lab**](#how-to-run-networkx-algorithms-in-memgraph-lab)
- [**How to implement custom NetworkX module**](#how-to-implement-custom-networkx-module)

## How to run NetworkX algorithms in Memgraph Lab

> NetworkX algorithms are integrated into Memgraph as query modules inside Memgraph’s open-source graph extension library [MAGE](/docs/mage). Head over to the guide on [how to call MAGE procedures](/docs/mage/usage/calling-procedures) to learn how to call all Memgraph procedures, including those that utilize the NetworkX library.

Reviewer: ```suggestion
NetworkX algorithms are integrated into Memgraph as query modules inside Memgraph’s open-source graph extension library [MAGE](/docs/mage). Head over to the guide on [how to call MAGE procedures](/docs/mage/usage/calling-procedures) to check how to call all Memgraph procedures, including those that utilize the NetworkX library.
```

> This how-to guide will show one simple example of calling a NetworkX procedure. 

Reviewer: ```suggestion
This how-to guide will show one simple example of calling a NetworkX procedure in Memgraph's visual interface Memgraph Lab. 
```

### 1. Connect to Memgraph

> To be able to try out this example on your own, you need a running Memgraph instance. Run Memgraph using the Memgraph Platform Docker image, which includes both the MAGE library and Memgraph Lab, a visual interface. 

Reviewer: ```suggestion
First, run Memgraph using the Memgraph Platform Docker image, which includes both the MAGE library and Memgraph Lab, a visual interface. 
```

To run the image, open a command-line interpreter and run the following Docker command:

```
docker run -it -p 7687:7687 -p 7444:7444 -p 3000:3000 memgraph/memgraph-platform:latest
```

[Connect to Memgraph](/docs/memgraph-lab/connect-to-memgraph#connecting-to-memgraph) via Memgraph Lab which is running at `localhost:3000`. 

Check out the [installation guide](/docs/memgraph/installation) for other installation options. If you wish to avoid the installation, you can also use [Memgraph Cloud](/docs/memgraph-cloud/).

### 2. Load the dataset

> First, head over to the **Datasets** section to load a dataset. Then, choose the **Europe backpacking dataset**. 

Reviewer: ```suggestion
Head over to the **Datasets** section to load a dataset and load the **Europe backpacking dataset**. 
```

<img src={require('../data/how-to-guides/nxalg-how-to-1.png').default} className={"imgBorder"}/>

### 3. Run NetworkX algorithm

Once the dataset is loaded, go to the **Query Modules** section and search for `nxalg` module. Click on the arrow next to the module name to **view module details**. 

<img src={require('../data/how-to-guides/nxalg-how-to-2.png').default} className={"imgBorder"}/>

The goal is to run the [`is_bipartite()`](/docs/mage/query-modules/python/nxalg#is_bipartite) procedure to check whether the graph is bipartite. 

<img src={require('../data/how-to-guides/nxalg-how-to-3.png').default} className={"imgBorder"}/>

> Copy the query, go to the **Query Execution** tab and paste the query into the **Cypher Editor**. If you need the query, here it is:

Reviewer: ```suggestion
Copy the query, go to the **Query Execution** tab and paste the query into the **Cypher Editor**:
```

```cypher 
CALL nxalg.is_bipartite() YIELD is_bipartite;
```

By clicking on the **Run Query** button, you can see that the Europe backpacking graph is not bipartite.

<img src={require('../data/how-to-guides/nxalg-how-to-4.png').default} className={"imgBorder"}/>

In the same way, you can run other procedures from the `nxalg` module and the procedures from the `graph_analyzer` module, which can be found in the **Query Modules** section.

> ## How to implement custom NetworkX module

Reviewer: ```suggestion
## How to implement custom NetworkX module in Memgraph Lab
```

Besides using already implemented modules, you can create your own module which utilizes the NetworkX library.
To learn how to implement a custom query module, head over to the [example of query module in Python](/docs/memgraph/reference-guide/query-modules/implement-custom-query-modules/custom-query-module-example#python-api).

> Since Memgraph is integrated with NetworkX, you can import NetworkX library inside your Python code and start coding. In this guide, we will create a new query module which utilizes the NetworkX library with the help of Memgraph Lab.

Reviewer: ```suggestion
Since Memgraph is integrated with NetworkX, you can import NetworkX library inside Python code. This guide will show you how to create a new query module that utilizes the NetworkX library within Memgraph Lab.
```

### 1. Connect to Memgraph

> To be able to try out this example on your own, you need a running Memgraph instance. Run Memgraph using the Memgraph Platform Docker image, which includes both the MAGE library and Memgraph Lab, a visual interface. 

Reviewer: ```suggestion
First, run Memgraph using the Memgraph Platform Docker image, which includes both the MAGE library and Memgraph Lab. 
```

To run the image, open a command-line interpreter and run the following Docker command:

```
docker run -it -p 7687:7687 -p 7444:7444 -p 3000:3000 memgraph/memgraph-platform:latest
```

[Connect to Memgraph](/docs/memgraph-lab/connect-to-memgraph#connecting-to-memgraph) via Memgraph Lab which is running at `localhost:3000`. 

Check out the [installation guide](/docs/memgraph/installation) for other installation options. If you wish to avoid the installation, you can also use [Memgraph Cloud](/docs/memgraph-cloud/).

### 2. Load the dataset

> First, load a dataset in the **Datasets** section by searching and loading the Karate club friendship network dataset. 

Reviewer: ```suggestion
In the **Datasets** section, find and load the Karate club friendship network dataset. 
```

<img src={require('../data/how-to-guides/nxalg-how-to-5.png').default} className={"imgBorder"}/>

### 3. Implement a custom query module

Once the dataset is loaded, go to the **Query modules** section. The goal is to create a community detection algorithm that can partition a network into multiple communities with the help of the NetworkX library. Click on the **New Module** and type in the module name, e.g., `communities`. 

<img src={require('../data/how-to-guides/nxalg-how-to-6.png').default} className={"imgBorder"}/>

There is a sample Python code on the next screen, inside the code editor. Select it, delete it and paste the following code:

> ```python

Reviewer: You don't want to explain the code a bit? Just in a few sentences, or make comments in the code?

Author: No. Just kidding. I will, thanks

Reviewer: No is a perfectly fine answer to me if you think it should be evident from the code itself :D
